ruby/object_oriented_programming_basics/managing_ruby_projects.md

```diff
@@ -0,0 +1,239 @@
+
+### Making use of multiple files
+
+If you are to split your code across multiple files, you first will need to know how to make sure code from one file can be used in another file. There are two main ways to do that:
+
+- `require_relative` which takes in a string of a relative path, like `require_relative 'lib/bogo_sort'`, is going to look for `lib/bogo_sort.rb` relative to the location of the current file.
+- `require` is trickier - there are two ways of using it: one is providing a relative path which is recognized by using `../` or `./` at the start or an absolute path. This works like `require_relative`. Another way is by providing a file name without the extension, like `require 'csv'` is going to look for a `csv.rb` in the Ruby's `$LOAD_PATH` global variable which by default contains the Ruby standard library. There are other file extensions it might look for, but this is not important at this point. In addition to that, if it doesn't find that file in `$LOAD_PATH` it is going to look through installed gems (more on those later) to see if the file is there.
+
+Both of those approaches are going to execute the file, allowing you to use their contents. If you try to require something for the second time, nothing will happen, and the requires will return `false`.
+
+Convention is that `require_relative` is used for your own code, while `require` is used for things outside of it, like gems that your app depend on. Now, there's couple of things that are important to know about requiring files in Ruby. <span id="namespace">First is that all required code is put into the same namespace. This means that if you have the same names for methods, modules, classes and so on they will be added together in order they were required.</span> For example:
+
+```rb
+# all files are in the same directory for simplicity's sake
+
+# foo.rb
+def foo(bar)
+  `#{bar} is awesome!`
+end
+
+# footoo.rb
+def foo(bar)
+  `#{bar} is awful!`
+end
+
+# my_cool_project.rb
+require_relative 'foo'
+require_relative 'footoo'
+
+puts foo('Cereal')
+#=> Cereal is awful!
+# Since foo is defined twice, the last definition wins out.
+```
+
+To make sure code doesn't get overwritten, Rubyists wrap their code in modules which give them the benefits of a namespace:
+
+```rb
+# all files are in the same directory for simplicity's sake
+
+# foo.rb
+module Foo
+  def self.foo(bar)
+    `#{bar} is awesome!`
+  end
+end
+# footoo.rb
+module Footoo
+  def self.foo(bar)
+    `#{bar} is awful!`
+  end
+end
+# my_cool_project.rb
+require_relative 'foo'
+require_relative 'footoo'
+
+puts Foo.foo('Cereal')
+#=> Cereal is awesome!
+puts Footoo.foo('Marmite')
+#=> Marmite is awful!
+puts foo('Cereal')
+#=> Errors out - there's no longer a free floating foo method to use.
+```
+
+Another benefit of this approach is that you don't need to hold all the code for part of your app in one file:
+
+```rb
```

Comment:
Sounds good. Since I want to keep the module-namespace example, there's no need to dive deeper into it and indeed a class example should be more understandable.